Add support for the official splicing protocol

t-bast opened the pull request:
We replace our experimental version of `splice_init`, `splice_ack` and `splice_locked` by their official version (see https://github.com/lightning/bolts/pull/1160). If our peer is using the experimental feature bit, we convert our outgoing messages to use the experimental encoding and incoming messages to the official messages.

We also change the TLV fields added to `tx_add_input`, `tx_signatures` and `splice_locked` to match the spec version. We always write both the official and experimental TLV to updated nodes (because the experimental one is odd and will be ignored) but we drop the official TLV if our peer is using the experimental feature, because they won't understand the even TLV field.

We do the same thing for the `commit_sig` TLV. For peers who support the official splicing version, we insert the `start_batch` message before the batch of `commit_sig` messages as specified in https://github.com/lightning/bolts/pull/1160.

We also support both the legacy and official reconnection logic (TLV fields in `channel_reestablish`). It is a bit painful because both protocols use TLV 1, but for different purposes, so we need to do it in a somewhat hacky way at the codec level.

This ensures that we're able to create splice transactions, even with non-upgraded nodes (full backwards-compatibility). It would be best to make sure Phoenix users upgrade to a version that uses the official protocol though, to allow cleaning up the codebase and guarantee that we haven't missed an edge case.

E2E tests:

- no issues making payments and reconnecting to lnd 0.20.1
- no issues making payments and reconnecting to cln 25.12.1
- no issues making payments and reconnecting to ldk with https://github.com/jkczyz/ldk-server/tree/2026-03-splicing-rbf (no issues with RBF either)

remyers commented:
This is an observation that occurred to me while reviewing this PR that might be relevant for a future PR ..

There will be times when using `CMD_BUMP_FUNDING_FEE` instead of `CMD_SPLICE` would cost less on-chain and confirm faster if you want to perform an additional splice while a pending splice is unconfirmed. The total cost to create a new funding transaction with `CMD_SPLICE` will often be higher than making the same changes with `CMD_BUMP_FUNDING_FEE` at a slightly higher fee rate. If you are blocked from making a new splice because a pending splice has been fee bumped, you could also make a new fee bump to splice in/out value rather than waiting for the previously rbf'd splice to confirm.

This seems to be technically possible in the protocol because CMD_BUMP_FUNDING_FEE uses the interactive tx protocol and can change the funding contribution with the `funding_output_contribution` TLV. Figuring out if a splice or rbf-splice is cheaper will also need to consider the additional fees paid to bump our counter party's inputs and outputs.

Perhaps the biggest downside I can see (beside complexity) is that if your pending splice or rbf confirms before a rbf-splice, then you would need to renegotiate it as a new splice.

Am I missing something here wrt fee savings? Do you think it would be worth considering this situation now rather than later?

t-bast commented:
> Perhaps the biggest downside I can see (beside complexity) is that if your pending splice or rbf confirms before a rbf-splice, then you would need to renegotiate it as a new splice.

Yes, the main reason we're not providing this yet (even though this is in theory possible) is to manage complexity. It can get really complex very quickly once you start going down that road, because if each splice "part" maps to a specific action (an on-the-fly funding), it can be a huge mess to reconcile which parts where actually done and which parts need to be replayed if a previous RBF attempt confirms.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:
```diff
@@ -885,68 +918,73 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
+            log.info("rejecting splice request: the previous splice has unconfirmed rbf attempts ({})", previousTxs.mkString(", "))
+            stay() using d.copy(spliceStatus = SpliceStatus.SpliceAborted) sending TxAbort(d.channelId, InvalidSpliceWithUnconfirmedRbf(d.channelId, previousTxs).getMessage)
           } else {
             log.info(s"accepting splice with remote.in.amount=${msg.fundingContribution} remote.in.push=${msg.pushAmount}")
```
If 0-conf is activated, should we still only allow a limited number of previous unconfirmed parent txs? and/or have some sort of rate limit on?  or perhaps it doesn't matter if we already trust our peer enough to activate 0-conf..

remyers reviewed (changes requested):
Looks good - the refactor for RBF makes it easier to read/understand.

I found a few nits but main issues to take a look at are related to adding logic to reject an TxInitRbf or SpliceInit when the parent splice is unconfirmed.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version4/ChannelCodecs4.scala:
```diff
@@ -658,7 +658,7 @@ private[channel] object ChannelCodecs4 {
         ("lastChecked" | blockHeight) ::
-        ("rbfStatus" | rbfStatusCodec) ::
+        ("status" | dualFundingStatusCodec) ::
         ("deferred" | optional(bool8, lengthDelimited(channelReadyCodec)))).as[DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED]
```
Ah! I see, it only actually reads/writes 0x01 or 0x02 in dualFundingStatusCodec to indicate if the state is waiting for sigs or waiting for confirmation.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala:
```diff
@@ -885,68 +918,73 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
+            val previousTxs = d.commitments.active.filter(_.fundingTxIndex == d.commitments.latest.fundingTxIndex).map(_.fundingTxId)
+            log.info("rejecting splice request: the previous splice has unconfirmed rbf attempts ({})", previousTxs.mkString(", "))
+            stay() using d.copy(spliceStatus = SpliceStatus.SpliceAborted) sending TxAbort(d.channelId, InvalidSpliceWithUnconfirmedRbf(d.channelId, previousTxs).getMessage)
           } else {
```
Looks good!

remyers reviewed (changes requested):
Changes to tests to disallow sequences of unconfirmed splices looks good. I only had a few minor nits/questions.

remyers commented on eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala:
```diff
@@ -1373,93 +1478,86 @@ class NormalSplicesStateSpec extends TestKitBaseClass with FixtureAnyFunSuiteLik
+    // From Alice's point of view, we now have two unconfirmed splices.
 
     alice2bob.ignoreMsg { case _: ChannelUpdate => true }
     bob2alice.ignoreMsg { case _: ChannelUpdate => true }
```
Ohh, I see. Its used before you do the disconnect/reconnect sequence.

t-bast committed:
Use official splice messages

We replace our experimental version of `splice_init`, `splice_ack` and
`splice_locked` by their official version. If our peer is using the
experimental feature bit, we convert our outgoing messages to use the
experimental encoding and incoming messages to the official messages.

We also change the TLV fields added to `tx_add_input`, `tx_signatures`
and `splice_locked` to match the spec version. We always write both the
official and experimental TLV to updated nodes (because the experimental
one is odd and will be ignored) but we drop the official TLV if our
peer is using the experimental feature, because it won't understand the
even TLV field.

We do the same thing for the `commit_sig` TLV. For peers who support the
official splicing version, we insert the `start_batch` message before
the batch of `commit_sig` messages.

This guarantees backwards-compatibility with peers who only support the
experimental feature.

t-bast committed:
Re-work retransmission on reestablish

We introduce a `retransmit_flags` field to `my_current_funding_locked`
and `next_funding` to ask our peer to retransmit `commitment_signed` or
`announcement_signatures` if we're expecting them. With this change, we
don't need to retransmit `splice_locked` on reconnection anymore to
trigger the exchange of `announcement_signatures`. We don't need to
retransmit it to let our peer know that we've seen enough confirmations
for the splice either, since `my_current_funding_locked` implies that.

This allows us to completely remove retransmission of `splice_locked` on
reconnection, and also get rid of the `your_last_funding_locked` TLV,
which greatly simplifies the reconnection logic. We do keep them for
backwards-compatibility with existing Phoenix users though, but we'll
be able to clean it up once they have updated.

Note that this works with taproot channels since we will simply provide
nonces in `channel_reestablish` when we need our peer to send
announcement signatures (not supported yet since taproot channels are
never announced).

We rollback using the `next_commitment_number` to let our peer know that
we haven't received their `commit_sig` and instead use the retransmit
flags added to the `next_funding` TLV, unless our peer is using the
legacy splicing protocol.

pm47 commented on eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala:
```diff
@@ -118,7 +118,7 @@ class OpenChannelInterceptorSpec extends ScalaTestWithActorTestKit(ConfigFactory
     import f._
 
-    val features = defaultFeatures.add(Features.SplicePrototype, FeatureSupport.Optional).add(Features.OnTheFlyFunding, FeatureSupport.Optional)
+    val features = defaultFeatures.add(Features.Splicing, FeatureSupport.Optional).add(Features.OnTheFlyFunding, FeatureSupport.Optional)
```
To ease rebase on feature branches:
```suggestion
    val features = defaultFeatures
      .add(Features.Splicing, FeatureSupport.Optional)
      .add(Features.OnTheFlyFunding, FeatureSupport.Optional)
```

t-bast replied:
Done in https://github.com/ACINQ/eclair/pull/2887/commits/668becb06a7b7dc3e19a22b8cae00de097f1ee62

pm47 commented on eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala:
```diff
@@ -210,7 +210,7 @@ class OpenChannelInterceptorSpec extends ScalaTestWithActorTestKit(ConfigFactory
     import f._
 
-    val features = defaultFeatures.add(Features.SplicePrototype, FeatureSupport.Optional).add(Features.OnTheFlyFunding, FeatureSupport.Optional)
+    val features = defaultFeatures.add(Features.Splicing, FeatureSupport.Optional).add(Features.OnTheFlyFunding, FeatureSupport.Optional)
```
```suggestion
    val features = defaultFeatures
      .add(Features.Splicing, FeatureSupport.Optional)
      .add(Features.OnTheFlyFunding, FeatureSupport.Optional)
```

t-bast replied:
Done in https://github.com/ACINQ/eclair/pull/2887/commits/668becb06a7b7dc3e19a22b8cae00de097f1ee62

t-bast committed:
Simplify rebase

pm47 commented on eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala:
```diff
@@ -143,14 +144,14 @@ case class TxSignatures(channelId: ByteVector32,
+      case Some(IndividualSignature(sig)) => Set(TxSignaturesTlv.PreviousFundingTxSig(sig), TxSignaturesTlv.ExperimentalPreviousFundingTxSig(sig))
       case Some(partialSig: PartialSignatureWithNonce) => Set(TxSignaturesTlv.PreviousFundingTxPartialSig(partialSig))
       case None => Set.empty
     }
```
```scala
  def apply(channelId: ByteVector32, 
            tx: Transaction, 
            witnesses: Seq[ScriptWitness], 
            previousFundingSig_opt: Option[ChannelSpendSignature]): TxSignatures = {
    val tlvs: Set[TxSignaturesTlv] = Set(
      previousFundingSig_opt match {
        case Some(IndividualSignature(sig)) => Set(TxSignaturesTlv.PreviousFundingTxSig(sig), TxSignaturesTlv.ExperimentalPreviousFundingTxSig(sig))
        case Some(partialSig: PartialSignatureWithNonce) => Set(TxSignaturesTlv.PreviousFundingTxPartialSig(partialSig))
        case None => Set.empty
      },
    ).flatten
    TxSignatures(channelId, tx.txid, witnesses, TlvStream(tlvs))
  }
```

t-bast replied:
I've done something slightly differently in https://github.com/ACINQ/eclair/pull/2887/commits/318ba314d5bc9824e031d07b772c84920ccfd6ff to flatten `Option`s instead of `Set`s, which should be even easier when rebasing.

t-bast committed:
Simplify rebase more
